README
NAME
       Encode::JIS2K - JIS X 0212 (aka JIS 2000) Encodings

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SYNOPSIS
         use Encode::JIS2K;
         use Encode qw/encode decode/;
         $euc_2k = encode("euc-jisx0213", $utf8);
         $utf8   = decode("euc-jisx0213", $euc_jp);

ABSTRACT
       This module implements encodings that covers JIS X 0213
       charset (AKA JIS 2000, hence the module name).  Encodings
       supported are as follows.

         Canonical     Alias                                      Description
         --------------------------------------------------------------------
         euc-jisx0213  qr/\beuc.*jp[ \-]?(?:2000|2k)$/i          EUC-JISX0213
                       qr/\bjp.*euc[ \-]?(2000|2k)$/i
                       qr/\bujis[ \-]?(?:2000|2k)$/i
         shiftjisx0213 qr/\bshift.*jis(?:2000|2k)$/i           Shift_JISX0213
                       qr/\bsjisp \-]?(?:2000|2k)$/i

         iso-2022-jp-3
         jis0213-1-raw                         JIS X 0213 plane 1, raw format
         jis0213-2-raw                         JIS X 0213 plane 2, raw format
         --------------------------------------------------------------------

DESCRIPTION
       To find out how to use this module in detail, see the
       Encode manpage.

what is JIS X 0213 anyway?
       Simply put, JIS X 0213 is a rework and reorganization of
       JIS X 0208 and JIS X 0212.  They consist of two 94x94
       planes which roughly corrensponds as follows;

         JIS X 0213 Plane 1 = JIS X 0208 + extension
         JIS X 0213 Plane 2 = JIS X 0212 reorganized + extension

       And here is the character repertoire there of at a glance.

                 # of codepoints     Kuten Ku (rows) used
         --------------------------------------------------------
         JIS X 0208         6,879    1..8,16..83
         JIS X 0213-1       8,762    1..94 (all!)
         JIS X 0212         6,067    2,6..7,9..11,16..77
         JIS X 0213-2       2,436    1,3..5,8,12..15,78..94
         -------------------------------------------------------
         (JIS X0213 Total) 11,197

       JIS X 0213 was designed to extend JIS X 0208 and JIS X
       0212 without being imcompatible to (classic) EUC-JP and
       Shift_JIS.  The following characteristics are as a result
       thereof.

       o JIS X plane 1 is (almost) a superset of JIS X 0208.
         However, with Unicode 3.2.0 the mappings differ in 3
         codepoints.

           Kuten   JIS X 0208 -> Unicode         JIS X 0213 -> Unicode
           --------------------------------------------------------------
           1-1-17  <UFFE3> # FULLWIDTH MACRON    <U203E> # OVERLINE
           1-1-29  <U2014> # EM DASH             <U2015> # HORIZONTAL BAR
           1-1-79  <UFFE5> # FULLWIDTH YEN SIGN  <U00A5> # YEN SIGN

       o By the same token, JIS X 0213 plane 2 contains JIS Dai-4
         Suijun Kanji (JIS Kanji Repertoire Level 4).  This
         allows EUC-JP's G3 to contain both JIS X 0212 and JIS
         0213 plane 2.

         However, JIS X 0212:1990 already contains many of Dai-4
         Suijun Kanji so EUC's G3 is subject to containing dupli-
         cate mappings.

       o Because of Halfwidth Katakana, Shift_JIS mapping has
         been tricky and it is even trickier.  Here is a regex
         that matches Shift_JISX0213 sequence (note: you have to
         "use bytes" to make it work!)

           $re_valid_shifjisx0213 =
             qr/^(?:
                  [\x00-\x7f] |                            # ASCII or
                  [\xa1-\xdf] |                            # JIS X 0201 KANA or
                  [\x81-\x9f\xe0-\xfc][\x40-\x7e\x80-\xfc] # JIS X 0213
                  )+$/xo;

       Note on EUC-JISX0213 (vs. EUC-JP)

       As of Encode-1.64, 'euc-jp' does support euc-jisx0213 for
       decoding.  However, 'euc-jp' in Encode and 'euc-jisx0213'
       differ as follows;

                           euc-jp                   euc-jisx0213
         --------------------------------------------------------------
         Decodes....       (0201-K|0208|0212|0213)  ditto
         Round-Trip  (|0)  (020-K|0208|0212)        JIS X (0201-K|0213)
         Decode Only (|3)  those only found in 0213
                                               those only found in 0212
         --------------------------------------------------------------

AUTHORS
       Dan Kogai <dankogai@dan.co.jp>

COPYRIGHT
       Copyright 2002 by Dan Kogai <dankogai@dan.co.jp>.

       This program is free software; you can redistribute it
       and/or modify it under the same terms as Perl itself.

       See http://www.perl.com/perl/misc/Artistic.html

SEE ALSO
       the Encode manpage, the Encode::JP manpage

       Japanese Graphic Character Set for Information Interchange
       -- Plane 1 http://www.itscj.ipsj.or.jp/ISO-IR/228.pdf

       Japanese Graphic Character Set for Information Interchange
       -- Plane 2 http://www.itscj.ipsj.or.jp/ISO-IR/229.pdf





Changes
#
# $Id: Changes,v 0.04 2023/11/15 03:55:42 dankogai Exp dankogai $
# Revision history for Perl extension Encode::JIS2K.
#
$Revision: 0.04 $ $Date: 2023/11/15 03:55:42 $
! JIS2K.pm README
  update POD and pod2text JIS2K.pm > README

0.04 2023/11/15 03:55:42
! JIS2K.pm
  resolved RT#150457: typo in documentation
  https://rt.cpan.org/Ticket/Display.html?id=150457
! t/JIS2K.t
  s/import Config/Config->import()/

0.03 2015/06/05 03:57:30
! Makefile.PL
  fix RT#67531 & RT#104967
  https://rt.cpan.org/Ticket/Display.html?id=67531
  https://rt.cpan.org/Ticket/Display.html?id=104967
! lib/Encode/JIS2K/2022JP3.pm
  fix RT#73401
  https://rt.cpan.org/Ticket/Display.html?id=73401

0.02 2005/05/12 07:00:10
! lib/Encode/JIS2K/2022JP3.pm
  find_encoding('iso-2022-jp-3')->name returns 'Canon', not 'iso-2022-jp-3'.
  Debugged by Hirata Keisuke tahi@gb3.so-net.ne.jp
  Message-Id: <20050512.154012.120210732.tahi@gb3.so-net.ne.jp>
0.01 2002/04/30 09:06:57 
  First Release



Makefile.PL
#
# $Id: Makefile.PL,v 0.1 2005/05/12 06:59:25 dankogai Exp $
#
use 5.7.2;
use strict;
use File::Spec;
use ExtUtils::MakeMaker;

# Please edit the following to the taste!
my $name = 'JIS2K';
my %tables = (
	     euc2k_t   => [ File::Spec->catfile(qw/ucm euc-jp2k.ucm/) ],
	     sjis2k_t  => [ File::Spec->catfile(qw/ucm sjis-2k.ucm/) ],
	     raw_t     => [ map { File::Spec->catfile('ucm', $_) }
			    qw/x0213-1.ucm x0213-2.ucm/
			    ],
	     );


# from perlvar
use Config;  # first appeared in 5.00307
my $secure_perl_path = $Config{perlpath};
if ($^O ne 'VMS') {
    my $secure_perl_path .= $Config{_exe}
        unless $secure_perl_path =~ m/$Config{_exe}$/i;
}
die "can't determine path to the perl I'm running under"
    unless ($secure_perl_path);
use File::Basename qw(dirname);
my $dpath = dirname $secure_perl_path;
my ($enc2xs, $encode_h) = ();
my $enc2xs_fpath = File::Spec->catfile($dpath, 'enc2xs');
$enc2xs = $enc2xs_fpath
    if (-x $enc2xs_fpath);
$enc2xs or die "enc2xs not found!";
print "enc2xs is $enc2xs\n";
my %encode_h = ();
for my $d (@INC){
    my $dir = File::Spec->catfile($d, "Encode");
    my $file =  File::Spec->catfile($dir, "encode.h");
    -f $file and $encode_h{$dir} = -M $file;
}
%encode_h or die "encode.h not found!";
# find the latest one
($encode_h) = sort {$encode_h{$b} <=> $encode_h{$a}} keys %encode_h;
print "encode.h is at $encode_h\n";

WriteMakefile(
              INC		=> "-I$encode_h",
	      NAME		=> 'Encode::'.$name,
	      VERSION_FROM	=> "$name.pm",
	      OBJECT		=> '$(O_FILES)',
	      'dist'		=> {
		  COMPRESS	=> 'gzip -9f',
		  SUFFIX	=> 'gz',
		  DIST_DEFAULT => 'all tardist',
	      },
	      MAN3PODS	=> {},
	      PREREQ_PM => {
			    'Encode'     => "1.41",
                           },
	      # OS 390 winges about line numbers > 64K ???
	      XSOPT => '-nolinenumbers',
	      );

package MY;

sub post_initialize
{
    my ($self) = @_;
    my %o;
    my $x = $self->{'OBJ_EXT'};
    # Add the table O_FILES
    foreach my $e (keys %tables)
    {
	$o{$e.$x} = 1;
    }
    $o{"$name$x"} = 1;
    $self->{'O_FILES'} = [sort keys %o];
    my @files = ("$name.xs");
    $self->{'C'} = ["$name.c"];
    # $self->{'H'} = [$self->catfile($self->updir,'encode.h')];
    my %xs;
    foreach my $table (sort keys %tables) {
	push (@{$self->{'C'}},"$table.c");
	# Do NOT add $table.h etc. to H_FILES unless we own up as to how they
	# get built.
	foreach my $ext (qw($(OBJ_EXT) .c .h .exh .fnm)) {
	    push (@files,$table.$ext);
	}
    }
    $self->{'XS'} = { "$name.xs" => "$name.c" };
    $self->{'clean'}{'FILES'} .= join(' ',@files);
    open(XS,">$name.xs") || die "Cannot open $name.xs:$!";
    print XS <<'END';
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#define U8 U8
#include "encode.h"
END
    foreach my $table (sort keys %tables) {
	print XS qq[#include "${table}.h"\n];
    }
    print XS <<"END";

static void
Encode_XSEncoding(pTHX_ encode_t *enc)
{
 dSP;
 HV *stash = gv_stashpv("Encode::XS", TRUE);
 SV *sv    = sv_bless(newRV_noinc(newSViv(PTR2IV(enc))),stash);
 int i = 0;
 PUSHMARK(sp);
 XPUSHs(sv);
 while (enc->name[i])
  {
   const char *name = enc->name[i++];
   XPUSHs(sv_2mortal(newSVpvn(name,strlen(name))));
  }
 PUTBACK;
 call_pv("Encode::define_encoding",G_DISCARD);
 SvREFCNT_dec(sv);
}

MODULE = Encode::$name	PACKAGE = Encode::$name
PROTOTYPES: DISABLE
BOOT:
{
END
    foreach my $table (sort keys %tables) {
	print XS qq[#include "${table}.exh"\n];
    }
    print XS "}\n";
    close(XS);
    return "# Built $name.xs\n\n";
}

sub postamble
{
    my $self = shift;
    my $dir  = "."; # $self->catdir('Encode');
    my $str  = "# $name\$(OBJ_EXT) depends on .h and .exh files not .c files - but all written by enc2xs\n";
    $str    .= "$name.c : $name.xs ";
    foreach my $table (sort keys %tables)
    {
	$str .= " $table.c";
    }
    $str .= "\n\n";
    $str .= "$name\$(OBJ_EXT) : $name.c\n\n";

    foreach my $table (sort keys %tables)
    {
	my $numlines = 1;
	my $lengthsofar = length($str);
	my $continuator = '';
	$str .= "$table.c : Makefile.PL";
	foreach my $file (@{$tables{$table}})
	{
	    $str .= $continuator.' '.$self->catfile($dir,$file);
	    if ( length($str)-$lengthsofar > 128*$numlines )
	    {
		$continuator .= " \\\n\t";
		$numlines++;
	    } else {
		$continuator = '';
	    }
	}
	my $plib   = $self->{PERL_CORE} ? '"-I$(PERL_LIB)"' : '';
	my $ucopts = '-"Q"';
	$str .=  
	    qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
	open (FILELIST, ">$table.fnm")
	    || die "Could not open $table.fnm: $!";
	foreach my $file (@{$tables{$table}})
	{
	    print FILELIST $self->catfile($dir,$file) . "\n";
	}
	close(FILELIST);
    }
    return $str;
}



lib/Encode/JIS2K/2022JP3.pm
package Encode::JIS2K::2022JP3;
use strict;

our $VERSION = do { my @r = (q$Revision: 0.2 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r };

use Encode qw(:fallbacks);

use base qw(Encode::Encoding);

my $Canon = 'iso-2022-jp-3';
$Encode::Encoding{$Canon} =
    bless {
          Name      =>   $Canon,
	   h2z       =>   1,
	   jis0212   =>   1,
	  } => __PACKAGE__;

# we override this to 1 so PerlIO works
sub needs_lines { 1 }

use Encode::CJKConstants qw(:all);

our $DEBUG = 0;

#
# decode is identical for all 2022 variants
#

sub decode($$;$)
{
    my ($obj, $str, $chk) = @_;
    my $residue = '';
    if ($chk){
	$str =~ s/([^\x00-\x7f].*)$//so;
	$1 and $residue = $1;
    }
    $residue .= jis_euc(\$str);
    $_[1] = $residue if $chk;
    return Encode::decode('euc-jisx0213', $str, FB_PERLQQ);
}

#
# encode is different
#

sub encode($$;$)
{
    require Encode::JP::H2Z;
    my ($obj, $utf8, $chk) = @_;
    # empty the input string in the stack so perlio is ok
    $_[1] = '' if $chk;
    my ($h2z, $jis0212) = @$obj{qw(h2z jis0212)};
    my $octet = Encode::encode('euc-jisx0213', $utf8, FB_PERLQQ) ;
    $h2z and &Encode::JP::H2Z::h2z(\$octet);
    euc_jis(\$octet, $jis0212);
    return $octet;
}



our $ESC_JISX0213_1 = "\e\$(O";
our $ESC_JISX0213_2 = "\e\$(P";

# JIS<->EUC

sub jis_euc {
    my $r_str = shift;
    $$r_str =~ s(
		 ($RE{JIS_0212}|$RE{JIS_0208}|$RE{ISO_ASC}|$RE{JIS_KANA})
		 ([^\e]*)
		 )
    {
	my ($esc, $chunk) = ($1, $2);
	if ($esc !~ /$RE{ISO_ASC}/o) {
	    $chunk =~ tr/\x21-\x7e/\xa1-\xfe/;
	    if ($esc =~ /$RE{JIS_KANA}/o) {
		$chunk =~ s/([\xa1-\xdf])/\x8e$1/og;
	    }
	    elsif ($esc =~ /$RE{JIS_0212}/o) {
		$chunk =~ s/([\xa1-\xfe][\xa1-\xfe])/\x8f$1/og;
	    }
	}
	$chunk;
    }geox;
    my ($residue) = ($$r_str =~ s/(\e.*)$//so);
    return $residue;
}

sub euc_jis{
    no warnings qw(uninitialized);
    my $r_str = shift;
    my $jis0212 = shift;
    $$r_str =~ s{
	((?:$RE{EUC_C})+|(?:$RE{EUC_KANA})+|(?:$RE{EUC_0212})+)
	}{
	    my $chunk = $1;
	    my $esc =
		( $chunk =~ tr/\x8E//d ) ? $ESC{KANA} :
		    ( $chunk =~ tr/\x8F//d ) ? $ESC_JISX0213_2 :
			$ESC_JISX0213_1;
	    if ($esc eq $ESC_JISX0213_2 && !$jis0212){
		# fallback to '?'
		$chunk =~ tr/\xA1-\xFE/\x3F/;
	    }else{
		$chunk =~ tr/\xA1-\xFE/\x21-\x7E/;
	    }
	    $esc . $chunk . $ESC{ASC};
	}geox;
    $$r_str =~
	s/\Q$ESC{ASC}\E
	    (\Q$ESC{KANA}\E|\Q$ESC_JISX0213_1\E|\Q$ESC_JISX0213_2\E)/$1/gox;
    $$r_str;
}

1;
__END__


=head1 NAME

Encode::JIS2K::2022JP3 -- internally used by Encode::JIS2K

=cut


JIS2K.pm
package Encode::JIS2K;
our $VERSION = sprintf "%d.%02d", q$Revision: 0.04 $ =~ /(\d+)/g;

use Encode;
use XSLoader;
XSLoader::load(__PACKAGE__,$VERSION);

use Encode::JIS2K::2022JP3;

Encode::define_alias(qr/\beuc.*jp[ \-]?(?:2000|2k)$/i => '"euc-jisx0213"');
Encode::define_alias(qr/\bjp.*euc[ \-]?(2000|2k)$/i   => '"euc-jisx0213"');
Encode::define_alias(qr/\bujis[ \-]?(?:2000|2k)$/i    => '"euc-jisx0213"');

Encode::define_alias(qr/\bshift.*jis(?:2000|2k)$/i    => '"shiftjisx0213"');
Encode::define_alias(qr/\bsjisp \-]?(?:2000|2k)$/i    => '"shiftjisx0213"');


1;
__END__

=head1 NAME

Encode::JIS2K - JIS X 0212 (aka JIS 2000) Encodings

=head1 SYNOPSIS

  use Encode::JIS2K;
  use Encode qw/encode decode/;
  $euc_2k = encode("euc-jisx0213", $utf8);
  $utf8   = decode("euc-jisx0213", $euc_jp);

=head1 ABSTRACT

This module implements encodings that covers JIS X 0213 charset (AKA
JIS 2000, hence the module name).  Encodings supported are as follows.

  Canonical     Alias                                      Description
  --------------------------------------------------------------------
  euc-jisx0213  qr/\beuc.*jp[ \-]?(?:2000|2k)$/i          EUC-JISX0213
                qr/\bjp.*euc[ \-]?(2000|2k)$/i
                qr/\bujis[ \-]?(?:2000|2k)$/i
  shiftjisx0213 qr/\bshift.*jis(?:2000|2k)$/i           Shift_JISX0213
                qr/\bsjisp \-]?(?:2000|2k)$/i

  iso-2022-jp-3
  jis0213-1-raw                         JIS X 0213 plane 1, raw format
  jis0213-2-raw                         JIS X 0213 plane 2, raw format
  --------------------------------------------------------------------

=head1 DESCRIPTION

To find out how to use this module in detail, see L<Encode>.

=head1 what is JIS X 0213 anyway?

Simply put, JIS X 0213 is a rework and reorganization of JIS X 0208
and JIS X 0212.  They consist of two 94x94 planes which roughly
corrensponds as follows;

  JIS X 0213 Plane 1 = JIS X 0208 + extension
  JIS X 0213 Plane 2 = JIS X 0212 reorganized + extension

And here is the character repertoire there of at a glance.

          # of codepoints     Kuten Ku (rows) used
  --------------------------------------------------------
  JIS X 0208         6,879    1..8,16..83
  JIS X 0213-1       8,762    1..94 (all!)
  JIS X 0212         6,067    2,6..7,9..11,16..77
  JIS X 0213-2       2,436    1,3..5,8,12..15,78..94
  -------------------------------------------------------
  (JIS X0213 Total) 11,197

JIS X 0213 was designed to extend JIS X 0208 and JIS X 0212 without
being imcompatible to (classic) EUC-JP and Shift_JIS.  The following
characteristics are as a result thereof.

=over 2

=item *

JIS X plane 1 is (almost) a superset of JIS X 0208.  However, with
Unicode 3.2.0 the mappings differ in 3 codepoints.

  Kuten   JIS X 0208 -> Unicode         JIS X 0213 -> Unicode
  --------------------------------------------------------------
  1-1-17  <UFFE3> # FULLWIDTH MACRON    <U203E> # OVERLINE
  1-1-29  <U2014> # EM DASH             <U2015> # HORIZONTAL BAR
  1-1-79  <UFFE5> # FULLWIDTH YEN SIGN  <U00A5> # YEN SIGN

=item *

By the same token, JIS X 0213 plane 2 contains JIS Dai-4 Suijun Kanji
(JIS Kanji Repertoire Level 4).  This allows EUC-JP's G3 to contain
both JIS X 0212 and JIS 0213 plane 2.

However, JIS X 0212:1990 already contains many of Dai-4 Suijun Kanji
so EUC's G3 is subject to containing duplicate mappings.

=item *

Because of Halfwidth Katakana, Shift_JIS mapping has been tricky and
it is even trickier.  Here is a regex that matches Shift_JISX0213
sequence (note: you have to "use bytes" to make it work!)

  $re_valid_shifjisx0213 =
    qr/^(?:
         [\x00-\x7f] |                            # ASCII or
	 [\xa1-\xdf] |                            # JIS X 0201 KANA or
         [\x81-\x9f\xe0-\xfc][\x40-\x7e\x80-\xfc] # JIS X 0213
         )+$/xo;

=back

=head2 Note on EUC-JISX0213 (vs. EUC-JP)

As of Encode-1.64, 'euc-jp' does support euc-jisx0213 for decoding.
However, 'euc-jp' in Encode and 'euc-jisx0213' differ as follows;

                    euc-jp                   euc-jisx0213
  --------------------------------------------------------------
  Decodes....       (0201-K|0208|0212|0213)  ditto
  Round-Trip  (|0)  (020-K|0208|0212)        JIS X (0201-K|0213)
  Decode Only (|3)  those only found in 0213
                                        those only found in 0212
  --------------------------------------------------------------

=head1 AUTHORS

Dan Kogai, C<< <dankogai+cpan at gmail.com> >>

=head1 COPYRIGHT

Copyright 2002-2023 by Dan Kogai C<< <dankogai+cpan at gmail.com> >>.

This program is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=head1 SEE ALSO

L<Encode>, L<Encode::JP>

Japanese Graphic Character Set for Information Interchange -- Plane 1
L<http://www.itscj.ipsj.or.jp/ISO-IR/228.pdf>

Japanese Graphic Character Set for Information Interchange -- Plane 2
L<http://www.itscj.ipsj.or.jp/ISO-IR/229.pdf>

=cut
